### PR TITLE
Launch elevated instances via shell:AppFolder

### LINKED ISSUE #14501: WindowsTerminal fails to launch elevate=true profiles on new install; elevate-shim does not set dllpath
### Windows Terminal version

1.15.2875.0

### Windows build number

10.0.22621.819

### Other Software

N/A - brand new system setup

### Steps to reproduce

This should repro on a brand new Windows 11 installation as long as you do not have vcruntime140.dll in your default system %PATH%.

- Install new Windows 11 System
- Install Windows Terminal from the App Store
- Launch Windows Terminal
- Create a new profile by duplicating the `Command Prompt` profile and setting the `elevate = true` flag
- Launch the new `Command Prompt (elevated)` profile (or whatever you named it)

Will provide additional details in the comments. I've got procmon traces, dump snapshots and a method to observe this w/o the failure.  Issue seems to be related to ShellExecuteExW in elevate-shim not setting the DllPath in the elevated process.

### Expected Behavior

New Windows Terminal window launches w/ elevated cmd.exe

### Actual Behavior

WindowsTerminal.exe fails to launch w/ "not found" errors for msvcp140.dll, vcruntime140.dll, vcruntime140_1.dll

image

image

image

### PR DESCRIPTION
**Pull request by jboelter**
This uses `shell:AppsFolder` to launch elevated instances of the app via
`ShellExecuteEx` and `runas` in elevate-shim.exe. The app to launch is
discovered via the `GetCurrentApplicationUserModelId` API.

e.g. `shell:AppsFolder\WindowsTerminalDev_8wekyb3d8bbwe!App`

This will fallback to launching `WindowsTerminal.exe` if it fails to
discover the app user model id to launch.

This also fixes a bug in elevate-shim where the first argument of
WinMain was lost (e.g. `new-tab`). 

Curiously, `AppLogic::RunAsUwp()` is never called and
`AppLogic::IsUwp()` is always false when running debug builds locally
(e.g. WindowsTerminalDev). It's not clear if this is an artifact of
development packages or something else.

## Validation Steps Performed

Various manual debug/execution scenarios.

Verified the fallback path by running the unbundled app by extracting
the `CascadiaPackage_0.0.1.0_x64.msix` from the 'drop' build artifact.

Fixes #14501

### CODE REVIEW

**Comment by jboelter** [comment]
> 

@microsoft-github-policy-service agree

**Comment by jboelter** [comment]
I'm not particularly attached to the specifics of the implementation; it was an iterative process to get here.

There may be a potential for (un)escaping issues in the command line conversion when it removes (replace w/ empty string) shell:AppsFolder\... from the cmdline args. i.e. arg0 from __wargv is not the same as what is found in the full cmd line string.

This is also assuming the string returned from `GetCurrentApplicationUserModelId` doesn't have spaces in it.

It also wasn't clear if terminal is always an AppX package, in which case the fallback to invoking WindowsTerminal.exe isn't necessary.

**Comment by jboelter** [comment]
@zadjii-msft 

Just had an idea while working on something entirely unrelated.  Turns out that `elevate-shim.exe` can successfully call `GetCurrentApplicationUserModelId`.  This means this change can be entirely isolated to the shim and not have to deal with any command line argument parsing.  It significant simplifies the shim to only a decision on how it generates the cmd to pass to shell execute.

I'll get an updated pull request out.

File: src/cascadia/ElevateShim/elevate-shim.cpp [hunk 1]
```diff
@@ -46,8 +99,10 @@ int __stdcall wWinMain(HINSTANCE, HINSTANCE, LPWSTR pCmdLine, int)
     seInfo.cbSize = sizeof(seInfo);
     seInfo.fMask = SEE_MASK_DEFAULT;
     seInfo.lpVerb = L"runas"; // This asks the shell to elevate the process
-    seInfo.lpFile = module.c_str(); // This is `...\WindowsTerminal.exe`
-    seInfo.lpParameters = pCmdLine; // This is `new-tab -p {guid}`
+    seInfo.lpFile = cmd.c_str(); // This is `shell:AppsFolder\...` or `...\WindowsTerminal.exe`
+    seInfo.lpParameters = GetCommandLine(); // This is `new-tab -p {guid}`
```
Comment (lhecker) [inline src/cascadia/ElevateShim/elevate-shim.cpp:103]: Why not use the previous `pCmdLine`?

**Review by lhecker — APPROVED** [review]

File: src/cascadia/ElevateShim/elevate-shim.cpp (hunk 1 above)
Comment (jboelter) [reply src/cascadia/ElevateShim/elevate-shim.cpp:103]: I found an innocuous bug in how the first parameter, `new-tab`, was dropped when passed to elevate-shim.exe by wWinMain given how it's launched by `CreateProcess`. It appears wWinMain is assuming param zero is the executable path even if it's not.

However, the elevate-shim process is launched with a command line that doesn't include the executable path.

https://github.com/microsoft/terminal/blob/8e041692b3321e2dd31106f4d9096673eb43b0e7/src/cascadia/TerminalApp/TerminalPage.cpp#L4126-L4135

---

This is the !peb from elevate-shim.exe when launched from a dev build of terminal as an appx (I setup Windbg Preview as the post-mortem debugger and put a DebugBreak() at the top of elevate-shim to break in easily). Notice the `CommandLine` is missing the executable filepath as the first argument. You can also see this in Process Monitor.

```
    CurrentDirectory:  'C:\windows\system32\'
    WindowTitle:  'C:\source\github.com.jboelter.terminal\src\cascadia\CascadiaPackage\bin\x64\Debug\AppX\elevate-shim.exe'
    ImageFile:    'C:\source\github.com.jboelter.terminal\src\cascadia\CascadiaPackage\bin\x64\Debug\AppX\elevate-shim.exe'
    CommandLine:  '  new-tab --profile "{0caa0dad-35be-5f56-a8ff-afceeeaa6101}"'
    DllPath:      'C:\source\github.com.jboelter.terminal\src\cascadia\CascadiaPackage\bin\x64\Debug\AppX;C:\Program Files\WindowsApps\Microsoft.VCLibs.140.00.Debug.UWPDesktop_14.0.30704.0_x64__8wekyb3d8bbwe;'
 ```

Comment (zadjii-msft) [reply src/cascadia/ElevateShim/elevate-shim.cpp:103]: Yeeeeeeeep that's totally on me. We probably could fix that down in TerminalPage.cpp, but 🤷

**Review by zadjii-msft — APPROVED** [review]
Every day I'm amazed and horrified. Let's give this a shot for 1.17. Somehow I have a feeling that this is going to be way more reliable, based on the fact that it's 500% crazier than what we had before 😆

**Review by DHowett — APPROVED** [review]
Thanks so much for doing this. It's clever and thoughtfully-implemented :smile:

**Comment by DHowett** [comment]
> Is `shell:AppsFolder` documented as an API? It's described in [Find the Application User Model ID of an installed app](https://learn.microsoft.com/en-us/windows/configuration/find-the-application-user-model-id-of-an-installed-app#to-find-the-aumid-by-using-file-explorer) for interactive use though, and there is `FOLDERID_AppsFolder` in [KNOWNFOLDERID](https://learn.microsoft.com/en-us/windows/win32/shell/knownfolderid).

I know that PowerToys uses it, but I don't know whether it's been publicly documented. That's something we should definitely pursue. Thank you :)

**Comment by AnrDaemon** [comment]
"Release notes" link in bot posts is 404'd.

**Pull request merged**